book-content/chapters/13-modules-scripts-declaration-files.md

```diff
@@ -0,0 +1,988 @@
+    searchAlbums(query: string): Promise<Album[]>;
+  };
+}
+```
+
+This almost works, except for the error. We can't use `declare` inside an ambient context: the `declare global` block is already ambient. So, we can remove the `declare` keyword:
+
+```typescript
+// inside musicUtils.ts
+
+declare global {
+  const ALBUM_API: {
+    getAlbumInfo(upc: string): Promise<Album>;
+    searchAlbums(query: string): Promise<Album[]>;
+  };
+}
+```
+
+Now the `ALBUM_API` variable has been put into the global scope.
+
+### `declare module`
+
+There are some situations where you need to declare types for a module that either doesn't have type definitions or is not included in the project directly.
+
+In these cases, you can use the `declare module` syntax to define types for the module.
+
+For example, say we are working with a `duration-utils` module that doesn't have type definitions.
+
+The first step would be to create a new file named `duration-utils.d.ts`. Then at the top of the file, the `declare module` syntax is used to define the types for the module:
+
+```typescript
+declare module "duration-utils" {
+  export function formatDuration(seconds: number): string;
+}
+```
+
+We use `export` to define what is being exported from the module.
+
+Like before, we are not including any implementation code in the `.d.ts` file– it's just the types that are being declared.
+
+Once the `duration-utils.d.ts` file is created, the module can be imported and used as usual:
+
+```typescript
+import { formatDuration, parseTrackData } from "music-utils";
+
+const formattedTime = formatDuration(309);
+```
+
+Just like normal declaration files, the types you add are not checked against the actual module - so it's important to keep them up to date.
+
+### Module Augmentation vs Module Overriding
+
+When using `declare module`, you can either augment an existing module or override it completely. Augmenting a module means appending new types to an existing module. Overriding a module means replacing the existing types with new ones.
+
+Choosing which you're doing depends on whether you're inside a module or a script.
+
+#### Inside a Module, `declare module` Augments
+
+If you're inside a module, `declare module` will augment the targeted module. For instance, you can add a new type to the `express` module:
+
+```typescript
+// inside express.d.ts
+declare module "express" {
+  export interface MyType {
+    hello: string;
+  }
+}
+
+export {}; // Adding an export turns this .d.ts file into a module
+```
+
+Now, across our project, we can import `MyType` from the `express` module:
+
+```typescript
+// anywhere.ts
+import { MyType } from "express";
+```
+
+We don't need to put this in a declaration file. We can get exactly the same behavior by changing `express.d.ts` to `express.ts`.
+
+This example is a little bit silly - there's no real use case for adding your own type to a module. But we'll see later that augmenting the types of modules can be extremely useful.
```

Comment:
It can be nice to include some types that were not exported by that module. It's surely a somewhat confusing hack but gives u some meaning and colocation so I can see how people could like it.